src/cascadia/TerminalConnection/ConhostConnection.cpp

```diff
@@ -234,16 +235,16 @@ namespace winrt::Microsoft::Terminal::TerminalConnection::implementation
             const BYTE* const endPtr{ buffer + dwRead };
             const BYTE* backIter{ endPtr - 1 };
             // If the last byte in the buffer was a byte belonging to a UTF-8 multi-byte character
-            if ((*backIter & 0b10000000) == 0b10000000)
+            if (WI_AreAllFlagsSet(*backIter & 0b10000000, 0b10000000))
             {
                 // Check only up to 3 last bytes, if no Lead Byte was found then the byte before must be the Lead Byte and no partials are in the buffer
                 for (DWORD dwSequenceLen{ 1UL }, stop{ dwRead < 4UL ? dwRead : 4UL }; dwSequenceLen < stop; ++dwSequenceLen, --backIter)
                 {
                     // If Lead Byte found
-                    if ((*backIter & 0b11000000) == 0b11000000)
+                    if (WI_AreAllFlagsSet(*backIter & 0b11000000, 0b11000000))
                     {
                         // If the Lead Byte indicates that the last bytes in the buffer is a partial UTF-8 code point then cache them
-                        if ((*backIter & bitmasks[dwSequenceLen]) != bitmasks[dwSequenceLen - 1])
+                        if (WI_IsAnyFlagClear(*backIter & bitmasks[dwSequenceLen], bitmasks[dwSequenceLen - 1]))

```

Comment:
This one deserves a comment -- since you're masking with the sequence prefix and _checking_ the bits from the next sequence prefix down, you may want to call that out specifically. It took me three reviews to figure out why you were doing this :smile:

I'm assuming it's because `0b11100000` is the prefix for a 3-byte sequence, but for it to be a 3-byte introducer those three high bits must be `110` (not `111`), and we're just using the off-by-one array index because those mask/check values are luckily adjacent?

Reply:
I begin with `dwSequenceLen = 1` which is the reason why `dwSequenceLen - 1` can't be off by one in the array.  
Actually only the `WI_IsAnyFlagClear(*backIter & 0b11000000, 0)` in the first iteration is tricky to understand. I checked before that a lead byte was found (that is, the two highest bits are 1). `WI_IsAnyFlagClear` will always return `true` here and the byte is getting cached.
The rest behaves exactly as you said. We need `0b111'00000` as mask to compare the result with `0b110'00000` etc. And that's the reason why naming these bytes would be ambiguous as I explained in the conversation above.

Reply:
@DHowett-MSFT Now that I read it twice I get this uncomfortable feeling. Should the first argument the second and vice versa?

Reply:
Hmm. 

`WI_IsAnyFlagClear(val, flags)` Any bit specified by `flags` is not set in `val`.

I would prefer the direct equality comparison here (which you had before without `WI_`). You want to make sure that _all bit values match within the mask._

Reply:
Alright. I figured something is going wrong here. Thanks!